operators/src/pro/cache/cache_tiles.rs

```diff
@@ -314,23 +309,354 @@ impl<T: Pixel> Stream for CacheTileStream<T> {
+                })));
+
+                // maybe the future is already finished, so we can try to return it immediately
+                let tile_future = state
+                    .as_mut()
+                    .as_pin_mut()
+                    .expect("The future must exist since we just set it");
+                // futures ready will return pending if the future is not ready yet
+                let tile_res_res = futures::ready!(tile_future.poll(cx));
+                state.set(None);
+                let Ok(tile_res) = tile_res_res else {
+                return std::task::Poll::Ready(Some(
+                    Err(CacheError::CouldNotDecompressElement),
+                ));
+            };
+                return std::task::Poll::Ready(Some(tile_res));
             }
         }
 
         std::task::Poll::Ready(None)
     }
 }
 
-pub enum TypedCacheTileStream {
-    U8(CacheTileStream<u8>),
-    U16(CacheTileStream<u16>),
-    U32(CacheTileStream<u32>),
-    U64(CacheTileStream<u64>),
-    I8(CacheTileStream<i8>),
-    I16(CacheTileStream<i16>),
-    I32(CacheTileStream<i32>),
-    I64(CacheTileStream<i64>),
-    F32(CacheTileStream<f32>),
-    F64(CacheTileStream<f64>),
+#[derive(Clone, Debug)]
+pub struct CompressedMaskedGrid<D, T> {
+    shape: D,
+    type_marker: PhantomData<T>,
+    data: Vec<u8>,
+    mask: Vec<u8>,
+}
+
+#[derive(Clone, Debug)]
+pub enum CompressedGridOrEmpty<D, T> {
+    Empty(EmptyGrid<D, T>),
+    Compressed(CompressedMaskedGrid<D, T>),
+}
+
+pub type CompressedRasterTile<D, T> = BaseTile<CompressedGridOrEmpty<D, T>>;
+pub type CompressedRasterTile2D<T> = CompressedRasterTile<GridShape2D, T>;
+
+fn unsafe_cast_data_slice_to_u8_slice<T>(data: &[T]) -> &[u8]
+where
+    T: Copy,
+{
+    unsafe { std::slice::from_raw_parts(data.as_ptr().cast::<u8>(), std::mem::size_of_val(data)) }
+}
+
+fn unsafe_cast_u8_slice_to_data_slice<T>(data: &[u8]) -> &[T]
+where
+    T: Copy,
+{
+    unsafe {
+        std::slice::from_raw_parts(
+            data.as_ptr().cast::<T>(),
+            data.len() / std::mem::size_of::<T>(),
+        )
+    }
+}
+
+impl<D, T> CompressedMaskedGrid<D, T> {
+    #[cfg(test)]
+    pub(crate) fn new(shape: D, data: Vec<u8>, mask: Vec<u8>) -> Self {
+        Self {
+            shape,
+            type_marker: PhantomData,
+            data,
+            mask,
+        }
+    }
+
+    pub fn compressed_data_size(&self) -> usize {
+        self.data.len()
+    }
+
+    pub fn compressed_mask_size(&self) -> usize {
+        self.mask.len()
+    }
+
+    pub fn compressed_size(&self) -> usize {
+        self.compressed_data_size() + self.compressed_mask_size()
+    }
+
+    pub fn compressed_data_slice(&self) -> &[u8] {
+        &self.data
+    }
+
+    pub fn compressed_mask_slice(&self) -> &[u8] {
+        &self.mask
+    }
+
+    pub fn shape(&self) -> &D {
+        &self.shape
+    }
+
+    pub fn compress_masked_grid<F>(grid: &MaskedGrid<D, T>, compression: F) -> Self
```

Comment:
would it make sense to make the compression a generic to statically bind the function? We don't switch between compressions anyway.

Reply:
[16f2480](https://github.com/geo-engine/geoengine/pull/848/commits/16f2480b1e57e0edbf8204191a6e23d696e324c7)

Reply:
shouldn't the `F: TileCompression` be part of the `CompressedMaskedGrid`. it doesn't really make sense to compress the tile with one compression and decompress it with another. IMHO you shouldn't have to know the compression algo when uncompressing a grid, but the grid should know which one is needed.